api/agent/call.go

```diff
@@ -182,6 +198,47 @@ func WithContext(ctx context.Context) CallOpt {
 	}
 }
 
+// This MUST run last
+func WithReservedSlot(ctx context.Context, slot Slot) CallOpt {
+	return func(a *agent, c *call) error {
+		// We need deadlines otherwise our slot will have expired in 1970... :)
+		now := time.Now()
+		slotDeadline := now.Add(time.Duration(c.Call.Timeout) * time.Second / 2)
+		execDeadline := now.Add(time.Duration(c.Call.Timeout) * time.Second)
+
+		c.slotDeadline = slotDeadline
+		c.execDeadline = execDeadline
+
+		StatsEnqueue(ctx)
+
+		a.startStateTrackers(ctx, c)
+		// The call to a.endStateTrackers() will then be done by Submit
+
+		// If we're provided a slot we use it, otherwise we go through the
+		// motions of allocating it.
+		if slot == nil {
+			// First check if we have capacity. We exit early if we don't, because
+			// reserving a slot must be a quick operation, as in, "can I reserve a
+			// slot RIGHT NOW please?".
+			if !a.resources.IsResourcePossible(c.Memory, uint64(c.CPUs), c.Type == models.TypeAsync) {
```

Comment:
I apologize, IsResourcePossible function name is not 100% clear here, but IsResourcePossible() checks if this request can ever be run. For example, if total RAM is 8GB, you can never ever run a 32GB job. So, here this logic seems flawed, as getSlot() will block for extended period even if IsResourcePossible() returned true.

An example:

8GB total ram, a request comes in for 1GB, IsResourcePossible() will return *always* True. And if there are other requests active (eg. 7.5 GB already in use), then getSlot() will block and wait until a free slot deadline available or context expires.

What you probably wanted was;

1) check IsResourcePossible() to fail impossibly big requests quickly (or skip this step and let '2' fail below)
2) run getSlot() with a quick/short slot deadline. Instead of Call.Timeout / 2.

Reply:
But getSlot() could also perform the download of the image, the container startup, etc, which will take time and fail the short deadline even if the node has all the capacity in the world.

What we need is just a capacity check. I guess we could use isResourceAvailableLocked() somehow.

Reply:
Actually the capacity check has to be more clever than that, because it has to recognize hot slots that can be reused... my bad, scratch my previous comment.

GetSlot() with a short deadline looks like a good compromise for now, except that it will probably fail if the image is not already on the machine - so in the 0-to-1 case nothing will ever run unless the image is preloaded and GetSlot() is fast.

I guess we are fighting the fact that the abstraction does not have the concept of a slot reservation (i.e. a ReserveSlot())... maybe we should just add that.

Reply:
I'm thinking about this and I think that the main problem here is that the slot creation and allocation mechanisms are based on a producer/consumer model which does not allow us to explicitly run queries like "give me a slot if you can, otherwise tell me why you can't".

The only thing the consumer (the call) can do is wait and hope that someday the producer (slot manager) will give them a slot. The consumer has no visibility of the current slot situation and can make no decisions to fail fast.

I'm open to suggestions, but I believe the current slot handling is unsuited to the "push" flow that we want to implement. Maybe we should have an alternative manager for handling slots...

Reply:
what are the benefits of reserving a slot in the LB if we have topological saturation knowledge? the latter is something of a requirement in some capacity or we risk O(n) placement. are there high % traffic cases where we would ask for a slot but not receive one (it seems low %)?

Reply:
Also reservation needs to take into account of some cases, such as idle and eject timers. Due to these, the "actual" agent can eject/kill and timeout containers (aka slots.) We've built into some logic into agent that can detect sleeping containers and kill them. The reservation must not be delayed enough for these to kick in, or some of these mechanisms need to be disabled.

An example;

1) For RouteA, LB_1 reserves a slot and hence a slot is waiting for the actual request.
2) RouteB arrives for another reservation (perhaps from LB_2), but system seems busy and this triggers eject logic and kills RouteA container.
3) LB_1 forwards RouteA actual exec request, but reservation is gone.

So, LB already needs to handle reservations that are no longer valid. And grpc calls/responses needs to be tuned to ensure they don't get hit by idle/eject timeouts too much.

LB seems to be too tied into Agent implementation with the reservation system. To make it work, we'll need tweaks in runner agent as well. Eg. Step 3 can in theory be retried, but today it can't without tweaks/fixes.

I think it would be good if we could make all this work without reserving slots.